### PR TITLE
[testsuite] exp script to build images for experiments

### PR DESCRIPTION
**Pull request by rustielin**
### Description

Create a script `exp` that allows blockchain devs to use the GHA workflows to build docker images remotely. This aids in experimentation with Forge. It does this by pushing local (commit) changes to a remote branch, triggers the workflow, and then deletes the remote branch, which is prefixed `exp/<ORIGIN_BRANCH_NAME>` so as to be human readable.

Note that local changes are not automatically committed, as the user will have to consciously do that. 

This saves costs since today blockchain devs are iterating on PRs, which has a nice UI and history, but also invokes all other GHA workflows like rust unit tests, etc, which adds up to the cost.

### Test Plan

Unit tests, and also run manually:

See the runs here: https://github.com/aptos-labs/aptos-core/actions/workflows/workflow-run-docker-rust-build.yaml

```
$ ./exp
[Mon, 12 Jun 2023 11:28:51] INFO [exp.try_push_new_branch:18] Branch exp/rustielin/exp already exists. Overriding its local state.
[Mon, 12 Jun 2023 11:28:51] INFO [exp.try_push_new_branch:22] Creating branch exp/rustielin/exp
[Mon, 12 Jun 2023 11:28:51] INFO [exp.try_push_new_branch:24] Pushing branch exp/rustielin/exp
[Mon, 12 Jun 2023 11:28:56] INFO [exp.try_push_new_branch:27] Successfully created new branch exp/rustielin/exp
[Mon, 12 Jun 2023 11:28:57] INFO [exp.main:90] GIT_SHA: 03baf970361e8df1e2ac2b5c5d37835da0320170
[Mon, 12 Jun 2023 11:28:57] INFO [exp.main:91] FEATURES: 
[Mon, 12 Jun 2023 11:28:57] INFO [exp.main:92] PROFILE: release
[Mon, 12 Jun 2023 11:28:57] INFO [exp.main:93] BUILD_ADDL_TESTING_IMAGES: false
[Mon, 12 Jun 2023 11:28:58] INFO [exp.main:114] Deleting remote branch exp/rustielin/exp
[Mon, 12 Jun 2023 11:29:00] INFO [exp.main:118] Workflow URL: https://github.com/aptos-labs/aptos-core/actions/workflows/workflow-run-docker-rust-build.yaml
[Mon, 12 Jun 2023 11:29:00] INFO [exp.main:120] Successfully triggered workflow
```

### CODE REVIEW

File: testsuite/test_framework/git.py [hunk 1]
```diff
@@ -15,3 +15,41 @@ def run(self, command) -> RunResult:
...
+
+    def is_clean(self) -> bool:
+        """Check if the git working directory is clean, using git status --porcelain"""
+        ret = self.run(["status", "--porcelain"])
+        succ = ret.succeeded()
+        out = ret.unwrap().decode().strip()
+        out_empty = len(out) == 0
+        return succ and out_empty
+
+    def remote_matches(self, remote: str, ref: str) -> bool:
```
Comment (perryjrandall) [inline testsuite/test_framework/git.py:33]: nit branch matches remote? this naming is a little unclear

File: testsuite/test_framework/git.py [hunk 2]
```diff
@@ -15,3 +15,41 @@ def run(self, command) -> RunResult:
...
+        """Check if the current branch matches the remote branch"""
+        # git ls-remote --heads  https://github.com/aptos-labs/aptos-core.git rustielin/exp
+        remote_commit_hash = self.get_remote_commit_hash(remote, ref)
+        local_commit_hash = self.get_commit_hash(ref)
+        return remote_commit_hash == local_commit_hash
+
+    def get_remote_commit_hash(self, remote: str, ref: str) -> str:
```
Comment (perryjrandall) [inline testsuite/test_framework/git.py:40]: you might wanna call this resolve_remote_ref or something like that, indicating that we're getting it from a branch

File: testsuite/test_framework/git.py [hunk 3]
```diff
@@ -15,3 +15,41 @@ def run(self, command) -> RunResult:
...
+        try:
+            return (
+                self.run(["ls-remote", "--heads", remote, ref])
+                .unwrap()
+                .decode()
+                .strip()
+                .split()[0]
+            )
+        except IndexError as e:
```
Comment (perryjrandall) [inline testsuite/test_framework/git.py:49]: I'm not a huge fan of exceptions for non exceptional behavior, i prefer LBYL in this case and or a return status like optional that can handle failure gracefully but thats my style nit

File: testsuite/test_framework/git.py [hunk 4]
```diff
@@ -15,3 +15,41 @@ def run(self, command) -> RunResult:
     def last(self, limit: int = 1) -> Generator[str, None, None]:
         for i in range(limit):
             yield self.run(["rev-parse", f"HEAD~{i}"]).unwrap().decode().strip()
+
+    def branch(self) -> str:
```
Comment (perryjrandall) [inline testsuite/test_framework/git.py:19]: You have many functions using rev-parse, can we consolidate them?

**Review by perryjrandall — APPROVED** [review]
nice tests!!!

all comments nits, awesome utility

**Comment by rustielin** [comment]
Added in place an option for `./exp --tail`, which will track the workflow status, which allows for things like forge to trigger after image build

**Comment by rustielin** [comment]
woot now we  can trigger Forge tests and reuse images:

```
# subsequent runs
$ poetry run ./exp run --wait
[Fri, 16 Jun 2023 15:28:27] INFO [exp.try_push_new_branch:53] Branch exp/rustielin/exp already exists. Overriding its local state.
[Fri, 16 Jun 2023 15:28:27] INFO [exp.try_push_new_branch:57] Creating branch exp/rustielin/exp
[Fri, 16 Jun 2023 15:28:27] INFO [exp.try_push_new_branch:59] Pushing branch exp/rustielin/exp
[Fri, 16 Jun 2023 15:28:28] INFO [exp.try_push_new_branch:62] Successfully created new branch exp/rustielin/exp
{
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "schemaVersion": 2,
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:09ab2a8540fa9a2b43e36a5dd663632070b8b067545dcaf17585ca953819a921",
      "size": 2012,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:3d25c7e39a7338264565f9e049d9305acb434917a60fc556140ff66bbe408321",
      "size": 567,
      "annotations": {
        "vnd.docker.reference.digest": "sha256:09ab2a8540fa9a2b43e36a5dd663632070b8b067545dcaf17585ca953819a921",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      }
    }
  ]
}[Fri, 16 Jun 2023 15:28:30] INFO [exp.run:336] !!!Image validator-testing:79e792d136aabe4876b00db74d7bba397c7a30bf already exists, using it...
```

**Review by sherry-x — APPROVED** [review]

**Pull request merged**